src/client/ByteBuffer.h

```diff
@@ -0,0 +1,453 @@
+    }
+
+    inline void putInt8_ts(const int8_t *buf, uint32_t len) {
+        if (!p_buffer_)
+            return;
+        checkSize(len);
+        memcpy(&p_buffer_[position_], buf, len);
+        position_ += len;
+    }
+
+    ByteBuffer &putBytes(const int8_t *buf, uint32_t len, uint32_t index) {
+        position_ = index;
+        for (uint32_t i = 0; i < len; i++)
+            append<int8_t>(buf[i]);
+
+        return *this;
+    }
+
+    ByteBuffer &putChar(char value) {
+        append<int8_t>(value);
+        return *this;
+    }
+
+    ByteBuffer &putChar(char value, uint32_t index) {
+        insert<int8_t>(value, index);
+        return *this;
+    }
+
+    ByteBuffer &putShort(uint16_t value) {
+        append<uint16_t>(value);
+        return *this;
+    }
+
+    ByteBuffer &putShort(uint16_t value, uint32_t index) {
+        insert<uint16_t>(value, index);
+        return *this;
+    }
+
+    ByteBuffer &putInt(uint32_t value) {
+        append<uint32_t>(value);
+        return *this;
+    }
+
+    ByteBuffer &putInt(uint32_t value, uint32_t index) {
+        insert<uint32_t>(value, index);
+        return *this;
+    }
+
+    ByteBuffer &putLong(uint64_t value) {
+        append<uint64_t>(value);
+        return *this;
+    }
+
+    ByteBuffer &putLong(uint64_t value, uint32_t index) {
+        insert<uint64_t>(value, index);
+        return *this;
+    }
+
+    ByteBuffer &putFloat(float value) {
+        append<float>(value);
+        return *this;
+    }
+
+    ByteBuffer &putFloat(float value, uint32_t index) {
+        insert<float>(value, index);
+        return *this;
+    }
+
+    ByteBuffer &putDouble(double value) {
+        append<double>(value);
+        return *this;
+    }
+
+    ByteBuffer &putDouble(double value, uint32_t index) {
+        insert<double>(value, index);
+        return *this;
+    }
+
+    // Read Methods
+    int8_t get() {
+        return read<int8_t>();
+    }
+
+    int8_t get(uint32_t index) const {
+        return read<int8_t>(index);
+    }
+
+    inline int8_t getInt8_t(uint32_t index) const {
+        return p_buffer_[index];
+    }
+
+    void getBytes(int8_t *buf, uint32_t len) {
+        if (!p_buffer_ || position_ + len > limit_)
+            return;
+
+        for (uint32_t i = 0; i < len; i++) {
+            buf[i] = p_buffer_[position_++];
+        }
+    }
+
+    void getBytes(uint32_t index, int8_t *buf, uint32_t len) const {
+        if (!p_buffer_ || index + len > limit_)
+            return;
+
+        uint32_t pos = index;
+        for (uint32_t i = 0; i < len; i++) {
+            buf[i] = p_buffer_[pos++];
+        }
+    }
+
+    char getChar() {
+        return read<int8_t>();
+    }
+
+    char getChar(uint32_t index) const {
+        return read<int8_t>(index);
+    }
+
+    uint16_t getShort() {
+        return read<uint16_t>();
+    }
+
+    uint16_t getShort(uint32_t index) const {
+        return read<uint16_t>(index);
+    }
+
+    uint32_t getInt() {
```

Comment:
it is better to add const for function

Reply:
Internally, this method updates the field `position_`.